plugin/route53/route53.go

```diff
@@ -15,98 +20,180 @@ import (
 	"github.com/miekg/dns"
 )
 
-// Route53 is a plugin that returns RR from AWS route53
+// Route53 is a plugin that returns RR from AWS route53.
 type Route53 struct {
 	Next plugin.Handler
 
-	zones  []string
-	keys   map[string]string
-	client route53iface.Route53API
+	zoneNames []string
+	client    route53iface.Route53API
+	upstream  *upstream.Upstream
+
+	zMu   sync.RWMutex
+	zones map[string]*zone
 }
 
-// ServeDNS implements the plugin.Handler interface.
-func (rr Route53) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
-	state := request.Request{W: w, Req: r}
-	qname := state.Name()
+type zone struct {
+	id string
+	z  *file.Zone
+}
 
-	zone := plugin.Zones(rr.zones).Matches(qname)
-	if zone == "" {
-		return plugin.NextOrFailure(rr.Name(), rr.Next, ctx, w, r)
+// New returns new *Route53.
+func New(
+	ctx context.Context,
+	c route53iface.Route53API,
+	keys map[string]string,
+	up *upstream.Upstream) (*Route53, error) {
+
+	zones := make(map[string]*zone, len(keys))
+	zoneNames := make([]string, 0, len(keys))
+	for dns, id := range keys {
+		_, err := c.ListHostedZonesByNameWithContext(ctx, &route53.ListHostedZonesByNameInput{
+			DNSName:      aws.String(dns),
+			HostedZoneId: aws.String(id),
+		})
+		if err != nil {
+			return nil, err
+		}
+		zones[dns] = &zone{id: id, z: file.NewZone(dns, "")}
+		zoneNames = append(zoneNames, dns)
 	}
+	return &Route53{
+		client:    c,
+		zoneNames: zoneNames,
+		zones:     zones,
+		upstream:  up,
+	}, nil
+}
 
-	output, err := rr.client.ListResourceRecordSets(&route53.ListResourceRecordSetsInput{
-		HostedZoneId:    aws.String(rr.keys[zone]),
-		StartRecordName: aws.String(qname),
-		StartRecordType: aws.String(state.Type()),
-		MaxItems:        aws.String("1"),
-	})
-	if err != nil {
-		return dns.RcodeServerFailure, err
+// Run executes first update, spins up an update forever-loop.
+// Returns error if first update fails.
+func (h *Route53) Run(ctx context.Context) error {
+	if err := h.updateZones(ctx); err != nil {
+		return err
 	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Infof("breaking out of Route53 update loop: %v", ctx.Err())
+				return
+			case <-time.After(1 * time.Minute):
+				if err := h.updateZones(ctx); err != nil && ctx.Err() == nil /* Don't log error if ctx expired. */ {
+					log.Errorf("failed to update zones: %v", err)
```

Comment:
We've settled an starting log lines with an uppercase `failed` -> `Failed` (same for the other log lines)